content/en/blog/2023/secure-apps-with-istio/index.md

```diff
@@ -0,0 +1,355 @@
+identity on the connection between the peers, your system would have to get the identity from the
+identity cache which could be outdated or incomplete.
+
+These Identity caches that map identity to workload IPs are not [ACID](https://en.wikipedia.org/wiki/ACID)
+(Atomicity, Consistency, Isolation, and Durability) and you want your security system to be applied
+to something with strong guarantees. Consider the following properties and questions you may want
+to ask yourself:
+
+- Staleness: How can a peer verify that an entry in the cache is 'current'?
+- Incompleteness: If there's a cache miss and the system fails to close the connection, does the
+network become unstable when it's only the cache 'synchronizer' that is failing?
+- What if something simply doesn't have an IP? For example, an AWS Lambda service doesn’t by
+default have a public IP.
+- Non-transactional: if you read the identity twice will you see the same value? If you are not
+careful in your access policy or auditing implementation this can cause real issues.
+- Who will guard the guards themselves? Are there established practices to protect
+the cache like CA has? What proof do you have that the cache has not been tampered with? Are you
+forced to reason about (and audit) the security of some complex infrastructure that is not your CA?
+
+Some of the above are worse than others. You can apply the **failing closed** principle but that does not solve all of the above.
+
+Identities are also used in enforcing access policies such as authorization policy, and these
+access policies are in the request path where your system has to make decisions fast to allow or
+deny the access. Whenever identities become mistaken, access policies could be bypassed without
+being detected or audited. For example, your identity cache may have your `checkout` pod’s prior
+allocated IP address associated as one of the `checkout` identities. If the `checkout` pod gets
+recycled and the same IP address is just allocated to one of the `frontend` pods, that `frontend` pod could have the `checkout`'s identity before the cache is updated, which could cause wrong access
+policies to be enforced.
+
+Let us illustrate the identity cache stale problem assuming the following large scale multi-cluster deployment:
+
+1. 100 clusters where each cluster has 100 nodes with 20 pods per node. The number of total pods is 200,000.
+1. 0.2% of pods are being churned at all times (rollout, restarts, recovery, node churn, ...), each churn is a 10 second window.
+1. 400 pods which are being churned are distributed to 10,000 nodes’ (caches) every 10 secs
+1. If the cache synchronizer stalls what % stale is the system after 5 minutes - potentially as high as 6%!
+
+Above assumes the cache synchronizer is in a steady state. If cache synchronizer has a brown-out it would affect its health-checking which increases churn rate, leading to cascading instability.
```

Comment:
```suggestion
The above example assumes the cache synchronizer is in a steady state. If the cache synchronizer has a brown-out it would affect its health-checking which increases churn rate, leading to cascading instability.
```